warehouse/oidc/services.py

```diff
@@ -148,10 +148,51 @@ def get_key(self, key_id):
                 tags=[f"provider:{self.provider}", f"key_id:{key_id}"],
             )
             return None
-        return PyJWK(keyset[key_id])
+        return jwt.PyJWK(keyset[key_id])
+
+    def _get_key_for_token(self, token):
+        """
+        Return a JWK suitable for verifying the given JWT.
+
+        The JWT is not verified at this point, and this step happens
+        prior to any verification.
+        """
+        unverified_header = jwt.get_unverified_header(token)
+        return self.get_key(unverified_header["kid"])
 
     def verify(self, token):
-        return NotImplemented
+        key = self._get_key_for_token(token)
+
+        try:
+            # NOTE: Many of the keyword arguments here are defaults, but we
+            # set them explicitly to assert the intended verification behavior.
+            valid_token = jwt.decode(
+                token,
+                key=key,
+                algorithms=["RS256"],
+                verify_signature=True,
+                # "require" only checks for the presence of these claims, not
+                # their validity. Each has a corresponding "verify_" kwarg
+                # that enforces their actual validity.
+                require=["iss", "iat", "nbf", "exp", "aud"],
+                verify_iss=True,
+                verify_iat=True,
+                verify_nbf=True,
+                verify_exp=True,
+                verify_aud=True,
+                issuer=self.issuer_url,
+                audience="pypi",
+                leeway=30,
```

Comment:
N.B.: `leeway` defines the permitted clock skew, i.e. +/- 30 seconds. I chose that number because it seems to be the recommended best practice for JWTs (higher numbers indicate a server problem).